### PR TITLE
reduce unnecessary rendering from state changes

### PR DESCRIPTION
**Pull request by christianvogt**
There are many instances in app where `mapStateToProps` produces a state object that cannot be shallow compared and therefore the components below are always re-rendered on _every_ state change. This is especially problematic with `flags` and `firehose`. This PR addresses a few areas that were affected by unnecessary re-renders when state changes that were irrelevant to the component changed. Easily observed every 10 seconds when the `updateTimestamps` action would be fired, the app content and masthead would re-render. Or when typing in the list page filter the entire app would re-render on each keystroke.

* `features.ts`: `connectToFlags` results in a state object `{ flags: boolean[] }` which cannot be shallow compared because `flags` is always a brand new array instance. Implemented `react-redux` `connect` `areStatePropsEqual` to perform a deep comparison of the array. Without this, every component that depends on flags is re-rendering on _every_ state change.
*  `firehose.jsx`: Reduced unnecessary re-renders by implementing `areStatesEqual` and `areStatePropsEqual` to perform lightweight checks. This isn't a perfect solution but far better then the status quo. Again, without this change, firehose would update its children on _every_ state change.
* `list.tsx`: `List#stateToProps` actually only depends on the `UI.listSorts` state but has a complex `stateToProps` return value and therefore added a simple implementation of `areStatesEqual` to limit the connect updates to only state changes that affect the outcome.
* Used `React.memo` in several places because when the parent component re-renders, we often unnecessarily are re-rendering large child components even though they won't change.
* Refactored the `app.jsx` to move the main app contents into a separate file: `app-contents.tsx` and made use of `React.memo`. Also made the new `AppContents` component pure and used `withRouter` to force a re-render since it depends on routing changes.

As part of the investigation I have enabled redux dev tools.

There's still a lot of room for improvement that can be tackled in other PRs. Overall we need to be more diligent when connecting to redux store to compute simple state that can be shallow compared and then to perform derived computations in the render functions.

Sample before and after of re-rendering when typing in filter text:

render-fix-before

render-fix-after

### CODE REVIEW

**Comment by spadgett** [comment]
AWS throttling
/retest

**Comment by spadgett** [comment]
Related to https://jira.coreos.com/browse/CONSOLE-1393

**Comment by spadgett** [comment]
The orphaned record sets should be cleaned up now.

/retest

File: frontend/public/components/app-contents.tsx [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  const statusPage = appendActiveNamespace('/status');
+  return <Redirect to={statusPage} />;
+});
+
+const LazyRoute = (props) => <Route {...props} render={(componentProps) => <AsyncComponent loader={props.loader} kind={props.kind} {...componentProps} />} />;
+
+// use `withRouter` to force a re-render when routes change since we are using React.memo
+const AppContents = withRouter(React.memo(() => (
```
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/public/components/app-contents.tsx]: Yes it's the only change. I could have kept it in `app.jsx` but that file was very large, dealing with multiple concerns, so it made sense to separate out into it's own file.

**Comment by spadgett** [comment]
@christianvogt Test failure looks legitimate. PV page shouldn't have a project dropdown.

> Kubernetes resource CRUD operations.PersistentVolume does not have a namespace dropdown on non-namespaced objects 0.04s

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_console/1558/pull-ci-openshift-console-master-e2e-aws-console/1318/artifacts/e2e-aws-console/gui_test_screenshots/bf21b83f9bdd0f3d2b8c730b6ec306ef.png

**Comment by spadgett** [comment]
/lgtm cancel
so the bot doesn't keep retesting

**Comment by christianvogt** [comment]
@spadgett updated PR to target master and addressed issue causing e2e failure.

**Pull request merged**